hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeUpgradeActions.java

```diff
@@ -0,0 +1,199 @@
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import javax.inject.Inject;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FreezeUpgradeActions {
+    private static final Logger log = LogManager.getLogger(FreezeUpgradeActions.class);
+
+    private static final String PREPARE_UPGRADE_DESC = "software";
+    private static final String TELEMETRY_UPGRADE_DESC = "telemetry";
+    private static final String MANUAL_REMEDIATION_ALERT = "Manual remediation may be necessary to avoid node ISS";
+
+    public static final String NOW_FROZEN_MARKER = "now_frozen.mf";
+    public static final String EXEC_IMMEDIATE_MARKER = "execute_immediate.mf";
+    public static final String EXEC_TELEMETRY_MARKER = "execute_telemetry.mf";
+    public static final String FREEZE_SCHEDULED_MARKER = "freeze_scheduled.mf";
+    public static final String FREEZE_ABORTED_MARKER = "freeze_aborted.mf";
+
+    public static final String MARK = "✓";
+
+    private final NetworkAdminServiceConfig adminServiceConfig;
+    private final WritableFreezeStore freezeStore;
+
+    @Inject
+    public FreezeUpgradeActions(
+            @NonNull final NetworkAdminServiceConfig adminServiceConfig,
+            @NonNull final WritableFreezeStore freezeStore) {
+        requireNonNull(adminServiceConfig);
+        requireNonNull(freezeStore);
+
+        this.adminServiceConfig = adminServiceConfig;
+        this.freezeStore = freezeStore;
+    }
+
+    public void externalizeFreezeIfUpgradePending() {
+        // TODO: call networkCtx.hasPreparedUpgrade() once available (issue #6201)
+        // final var isUpgradePrepared = networkCtx.hasPreparedUpgrade();
+        final boolean isUpgradePrepared = true;
+
+        if (isUpgradePrepared) {
+            writeCheckMarker(NOW_FROZEN_MARKER);
+        }
+    }
+
+    public CompletableFuture<Void> extractTelemetryUpgrade(
+            @NonNull final byte[] archiveData, @Nullable final Instant now) {
+        requireNonNull(archiveData);
+        return extractNow(archiveData, TELEMETRY_UPGRADE_DESC, EXEC_TELEMETRY_MARKER, now);
+    }
+
+    public CompletableFuture<Void> extractSoftwareUpgrade(@NonNull final byte[] archiveData) {
+        requireNonNull(archiveData);
+        return extractNow(archiveData, PREPARE_UPGRADE_DESC, EXEC_IMMEDIATE_MARKER, null);
+    }
+
+    public void scheduleFreezeOnlyAt(@NonNull final Instant freezeTime) {
+        requireNonNull(freezeTime);
+        withNonNullDualState("schedule freeze", ds -> ds.freezeTime(freezeTime));
+    }
+
+    public void scheduleFreezeUpgradeAt(@NonNull final Instant freezeTime) {
```

Comment:
As these are all used in a switch expression, it's usually best to have them all return the same type (so the switch expression type is consistent).
You could have the ones that execute immediately (like this method) just return `CompletableFuture.completedFuture((Void)null)` to ensure they're all compatible.  A completed future will return immediately from any call to `get()` or `join()`.

Reply:
Fixed